### PR TITLE
Integrate PHPickerViewController in Media library

### PR DESCRIPTION
**Pull request by kean**
Part of https://github.com/wordpress-mobile/WordPress-iOS/issues/21190.

The changes affect both the WP and the Jetpack app. 

Screenshot 2023-08-16 at 10 39 52 AM

## Tests

> [!IMPORTANT]
> - Enable the "Native Photo Picker" feature flag before testing
> - The new picker was added **only** to the Media library screen itself

### Permissions

- Install a fresh build to reset the permissions
- Verify that to upload media, the app no longer asks for permissions

### Media Formats

- Verify that .heic images are uploaded as .jpeg (matches the existing production behavior of using a high-compatibility format by default)
- Verify that .webp images are uploaded as .heic (the existing production version doesn't support them; more info in the code comments)
- Verify that .gif are uploaded (with no modifications)
- Verify that video files are uploaded (and transcoded to .mp4)

You can add a WebP image from https://developers.google.com/speed/webp/gallery1: open it in the browser, long-tap and select “Save to Photos”.

### Image Resizing

- Go to Me / App Settings
- Set Max Image Upload Size to 200x200
- Upload an image
- Open the image details and verify that the resolution is lower than 200x200

### Removing GPS Data

- Enable "Remove Location From Media" in App Settings
- Upload an image with GPS data
- Verify that the GPS data was removed

You can use [test-image-device-photo-gps.jpg](https://github.com/wordpress-mobile/WordPress-iOS/blob/trunk/WordPress/WordPressTest/Test%20Data/test-image-device-photo-gps.jpg) for testing and macOS `Cmd+I` command to view photo EXIF.

### iCloud

- Open Jetpack on device A with "Optimize iPhone Storage" enabled
- Open the new photo picker
- Upload a large photo on device B
- Verify that the photo is automatically added to the picker on device A
- Select the photo before it gets fully downloaded
- Verify that the photo got uploaded to Jetpack

### Misc

- Verify that progress is updated during export and upload
- Verify that the selection is unlimited (as much as the system can chew)
- There should be no need to test the picker itself because it's fully native, but some of its advantages are outlined here https://github.com/wordpress-mobile/WordPress-iOS/issues/21190
- Verify that the picker uses the app's tint color

## Regression Notes
1. Potential unintended areas of impact: Media uploads
2. What I did to test those areas of impact (or what existing automated tests I relied on): manual and unit tests
3. What automated tests I added (or what prevented me from doing so): yes

PR submission checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding unit tests for my changes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

UI Changes testing checklist:
- [x] Portrait and landscape orientations.
- [x] Light and dark modes.
- [x] Fonts: Larger, smaller and bold text.
- [ ] High contrast.
- [ ] VoiceOver.
- [ ] Languages with large words or with letters/accents not frequently used in English.
- [ ] Right-to-left languages. (Even if translation isn’t complete, formatting should still respect the right-to-left layout)
- [x] iPhone and iPad. 
- [ ] Multi-tasking: Split view and Slide over. (iPad)

### CODE REVIEW

File: WordPress/Classes/Utility/Media/ItemProviderMediaExporter.swift [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+    var imageOptions: MediaImageExporter.Options?
+    var videoOptions: MediaVideoExporter.Options?
+
+    private let provider: NSItemProvider
+
+    init(provider: NSItemProvider) {
+        self.provider = provider
+    }
+
+    func export(onCompletion: @escaping (MediaExport) -> Void, onError: @escaping (MediaExportError) -> Void) -> Progress {
```
Comment (kean) [inline WordPress/Classes/Utility/Media/ItemProviderMediaExporter.swift:16]: It works nearly identical to the existing `MediaAssetExporter` (`PHAsset` support) and shares some of the code with it. 

Great post on `NSItemProvider`: https://www.humancode.us/2023/07/08/all-about-nsitemprovider.html.

File: WordPress/Classes/Utility/Media/ItemProviderMediaExporter.swift [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+    private static let supportedImageTypes: Set<String> = Set([
+        UTType.png,
+        UTType.jpeg,
+        UTType.gif,
+        UTType.heic,
+        UTType.svg
+    ].map(\.identifier))
+
+    private func hasConformingType(_ type: UTType) -> Bool {
+        provider.hasItemConformingToTypeIdentifier(type.identifier)
```
Comment (kean) [inline WordPress/Classes/Utility/Media/ItemProviderMediaExporter.swift:101]: There is a nice UTType-based API but it’s available only in iOS 16 :(

File: WordPress/WordPressTest/ItemProviderMediaExporterTests.swift [hunk 3]
```diff
@@ -0,0 +1,153 @@
+import Foundation
+import OHHTTPStubs
+import XCTest
+@testable import WordPress
+
+final class ItemProviderMediaExporterTests: XCTestCase {
```
Comment (kean) [inline WordPress/WordPressTest/ItemProviderMediaExporterTests.swift:6]: `MediaImageExporterTests` that’s used internally already has a bunch of tests, so we mostly rely on them and just test the integration.

**Comment by kean** [comment]
I've been running some additional tests and encountered a limitation with RAW files. I added a RAW file to the Photos library (6000×4000px) and tried uploading it, but the file provider serves the downscaled version:

```
(lldb) po CGImageSourceCreateImageAtIndex(source, 0, nil)
▿ Optional<CGImageRef>
  - some : <CGImage 0x10ba415d0> (IP)
	<<CGColorSpace 0x10a384460> (kCGColorSpaceICCBased; kCGColorSpaceModelRGB; sRGB IEC61966-2.1)>
		width = 1616, height = 1080, bpc = 8, bpp = 32, row bytes = 6464 
		kCGImageAlphaNoneSkipLast | 0 (default byte order)  | kCGImagePixelFormatPacked 
		is mask? No, has masking color? No, has soft mask? No, has matte? No, should interpolate? Yes
```

I tried uploading an even larger JPEG file (7200 × 5400px), and it worked perfectly and uploaded the full-resolution image. 

I have yet to find any documentation covering RAW support, so I'm asking around. There is a workaround: `PHPickerResult` contains an asset identifier, which you can always you to fetch the original data, but the problem is it requires access to the Photos library, which we are trying to avoid.

I'm not sure how important that scenario is, but it is still worth mentioning. WordPress doesn't [support RAW files](https://wordpress.com/support/accepted-filetypes/), so the app still has to transcode the image to JPEG. But the whole point of RAW files is to allow the users to pull out as many details from the photo as possible and create the JPEG they want. It's just not a thing that someone would want to upload a RAW file to WordPress and use the default iOS processing with no adjustments to the original photo.

**Comment by guarani** [comment]
Finished manual testing and this is looking great! I have a couple of questions I left below.

I'll go through the code separately and leave a review after that.

- [x] Permissions
- [x] Media Formats
	- [x] .heic 
		- ⚠️ On the iPhone simulator, the sample .heic image isn't being added to the WP.com media library (see video). However, .heic images on a physical device upload fine. I used the default .heic image that comes with the simulator (pink flowers). Do you see this too?
	- [x] .webp
		- Side note: The testing instructions say this doesn't work in production but it seems to work for me on 23.0 beta. 
		- Side note: I think there's a typo in the testing instructions, it should say uploaded as ".jpg"
	- [x] .gif 
	- [x] .mp4
		- Tested with .mov video on simulator and hevc on a physical device
- [x] Image Resizing
- [x] Removing GPS Data
	- ⚠️ Note that the native picker says "location included" which could be confusing given that the app's setting for this is OFF. I couldn't find a way to hide this in [the docs](https://developer.apple.com/documentation/photokit/phpickercapabilities). The previous interface didn't show this, so while I think there's some potential for user confusion, it isn't a blocker.
- [x] iCloud
	- Tested by adding a video to a shared library from another device and verified the native picker auto-updates while open to include the video. After adding it to the WP.com media library, I think it downloaded the full-resolution version from iCloud and then uploaded it to WP.com 👍
- [x] Misc
	- [x] Progress indicators (what is export)
		- ❓What does export refer to?
	- [x] Verify selection is unlimited
		- Tested adding 29 items 
	- [x] Verify app's tint color is used
		- This works well except for the picker's Options screen navigation bar background color, which is transparent and should be a solid white. Not sure if this is the bar appearance background color or a tint color (see below screenshot)


#### .heic image not uploading on the sim

https://github.com/wordpress-mobile/WordPress-iOS/assets/1898325/f5c93efe-5aeb-4f4f-b06c-071c05b2beee

#### Picker saying the location is included 

| This branch | 23.0 beta |
| - | - |
|  |  |

#### Picker navigation bar transparency color 

|  |
| - |

**Comment by kean** [comment]
> On the iPhone simulator, the sample .heic image isn't being added to the WP.com media library 

It's a [known issue](https://stackoverflow.com/questions/74244758/getting-a-pamediaconversionserviceerrordomain-error-2-when-importing-an-image-us) with the flower image on the simulator. The item provider fails to export it on sims; nothing we can do. It's shame it's the very first one in the list 😃

**Comment by kean** [comment]
> Note that the native picker says "location included" which could be confusing given that the app's setting for this is OFF.

The export options are a [new feature](https://9to5mac.com/2023/08/10/limit-iphone-photos-app-access-with-ios-17/) added in iOS 17. Is that what you are testing with?

There is no way to remove these options, but I think it's a great opportunity for us to remove some of the work that [MediaImageExporter](https://github.com/wordpress-mobile/WordPress-iOS/blob/5ee5cea73d2cdda811ab6df3b76f357de0661aa2/WordPress/Classes/Utility/Media/MediaImageExporter.swift#L275) does. I kept it as is, so it fully matched the production version, but it seems problematic in a few ways:

- Why is it using `CGImageSourceCreateThumbnailAtIndex` and not `CGImageSourceCreateImageAtIndex`?
- Why is it changing the compression ratio? Does it mean we lose signal when re-exporting the photos?

I think, by default, the app should upload the image as is with no modifications whatsoever. Suppose we remove the "remove GPS data" feature from the Jetpack app in favor of the built-in feature on iOS 17. In that case, we can re-evaluate the decisions in `MediaImageExporter` too. It'll be great to simplify it to reduce the amount of work the app does and also ensure we don't lose signal.

Btw, there is a screenshot attached to the message that says 23.0 beta, but the new photo picker isn't included in it.

> What does export refer to?

Exporting the item from `NSItemProvider` and processing it using `MediaImageExporter`. It takes a little under a second per a reasonably large photo on modern devices (which is way too much).

**Comment by guarani** [comment]
> It's a [known issue](https://stackoverflow.com/questions/74244758/getting-a-pamediaconversionserviceerrordomain-error-2-when-importing-an-image-us) with the flower image on the simulator. The item provider fails to export it on sims; nothing we can do. It's shame it's the very first one in the list 😃

Aha good catch, I didn't know that – it used to work at some point in the past.

> The export options are a [new feature](https://9to5mac.com/2023/08/10/limit-iphone-photos-app-access-with-ios-17/) added in iOS 17. Is that what you are testing with?

Yep, I'm on iOS 17 on my iPhone.

> There is no way to remove these options, but I think it's a great opportunity for us to remove some of the work that [MediaImageExporter](https://github.com/wordpress-mobile/WordPress-iOS/blob/5ee5cea73d2cdda811ab6df3b76f357de0661aa2/WordPress/Classes/Utility/Media/MediaImageExporter.swift#L275) does. I kept it as is, so it fully matched the production version, but it seems problematic in a few ways:
> 
> * Why is it using `CGImageSourceCreateThumbnailAtIndex` and not `CGImageSourceCreateImageAtIndex`?
> * Why is it changing the compression ratio? Does it mean we lose signal when re-exporting the photos?
> 
> I think, by default, the app should upload the image as is with no modifications whatsoever. Suppose we remove the "remove GPS data" feature from the Jetpack app in favor of the built-in feature on iOS 17. In that case, we can re-evaluate the decisions in `MediaImageExporter` too. It'll be great to simplify it to reduce the amount of work the app does and also ensure we don't lose signal.

I'm in favor of relying on the system "remove GPS data" if there's a way to ensure users don't accidentally upload GPS data by not realizing they need to manually turn it off at the system level.

> Btw, there is a screenshot attached to the message that says 23.0 beta, but the new photo picker isn't included in it.

The screenshot shows the system picker that's nested inside `MediaPicker-iOS` when the user selects more photos to grant the app access to. Here's a video showing it:

https://github.com/wordpress-mobile/WordPress-iOS/assets/1898325/ab9127c7-b24b-4843-a69d-20278f75991b

> > What does export refer to?
> 
> Exporting the item from `NSItemProvider` and processing it using `MediaImageExporter`. It takes a little under a second per a reasonably large photo on modern devices (which is way too much).

Gotcha 👍

**Comment by kean** [comment]
> I'm in favor of relying on the system "remove GPS data" if there's a way to ensure users don't accidentally upload GPS data by not realizing they need to manually turn it off at the system level

We can keep it in the app but avoid doing any extra processing if it's not there. Reading EXIF and verifying if it's there is a 
nearly instantaneous operation. Then we'd only need to check if the image size is greater than the preferred upload size and, again, avoid doing any extra processing if it's OK.

> The screenshot shows the system picker that's nested inside MediaPicker-iOS when the user selects more photos to grant the app access to. 

Ah, got it. Yes, it makes sense it doesn't have the export options: it doesn't export the photos.

**Comment by mokagio** [comment]
I'm going to bump this to the next release because we'll be code freezing [23.1](https://github.com/wordpress-mobile/WordPress-iOS/milestone/255) today and this hasn't been approved yet.

If this cannot wait two weeks and it's important that it makes it into this release, let me know and we'll organize a new beta once ready.

File: WordPress/Classes/ViewRelated/Media/NSIterProvider+Exportable.swift
Comment (guarani) [inline WordPress/Classes/ViewRelated/Media/NSIterProvider+Exportable.swift:1]: There's a typo in the filename _NSIter_

File: WordPress/Classes/ViewRelated/Media/PHPickerContoller+Extensions.swift
Comment (guarani) [inline WordPress/Classes/ViewRelated/Media/PHPickerContoller+Extensions.swift:1]: Typo in name _PHPickerContoller_

**Review by guarani — APPROVED** [review]
This looks good @kean! Tested in my last review 👍 The typos aren't blockers if you can fix them in a follow-up PR.

I'm excited to see the new image picker and I know this will fix a bunch of issues like https://github.com/wordpress-mobile/WordPress-iOS/issues/19576 once it's integrated across the rest of the app.

File: WordPress/Classes/ViewRelated/Media/PHPickerContoller+Extensions.swift
Comment (kean) [reply WordPress/Classes/ViewRelated/Media/PHPickerContoller+Extensions.swift:1]: Fixed

File: WordPress/Classes/ViewRelated/Media/NSIterProvider+Exportable.swift
Comment (kean) [reply WordPress/Classes/ViewRelated/Media/NSIterProvider+Exportable.swift:1]: Fixed

**Pull request merged**